data/abi/abi_type.go

```diff
@@ -0,0 +1,491 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package abi
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+/*
+   ABI-Types: uint<N>: An N-bit unsigned integer (8 <= N <= 512 and N % 8 = 0).
+            | byte (alias for uint8)
+            | ufixed <N> x <M> (8 <= N <= 512, N % 8 = 0, and 0 < M <= 160)
+            | bool
+            | address (alias for byte[32])
+            | <type> [<N>]
+            | <type> []
+            | string
+            | (T1, ..., Tn)
+*/
+
+// BaseType is an type-alias for uint32. A BaseType value indicates the type of an ABI value.
+type BaseType uint32
+
+const (
+	// Uint is the index (0) for `Uint` type in ABI encoding.
+	Uint BaseType = iota
+	// Byte is the index (1) for `Byte` type in ABI encoding.
+	Byte
+	// Ufixed is the index (2) for `UFixed` type in ABI encoding.
+	Ufixed
+	// Bool is the index (3) for `Bool` type in ABI encoding.
+	Bool
+	// ArrayStatic is the index (4) for static length array (<type>[length]) type in ABI encoding.
+	ArrayStatic
+	// Address is the index (5) for `Address` type in ABI encoding (an type alias of Byte[32]).
+	Address
+	// ArrayDynamic is the index (6) for dynamic length array (<type>[]) type in ABI encoding.
+	ArrayDynamic
+	// String is the index (7) for `String` type in ABI encoding (an type alias of Byte[]).
+	String
+	// Tuple is the index (8) for tuple `(<type 0>, ..., <type k>)` in ABI encoding.
+	Tuple
+)
+
+// Type is the struct that stores information about an ABI value's type.
+type Type struct {
+	abiTypeID  BaseType
+	childTypes []Type
+
+	// only can be applied to `uint` bitSize <N> or `ufixed` bitSize <N>
+	bitSize uint16
+	// only can be applied to `ufixed` precision <M>
+	precision uint16
+
+	// length for static array / tuple
+	/*
+		by ABI spec, len over binary array returns number of bytes
+		the type is uint16, which allows for only lenth in [0, 2^16 - 1]
+		representation of static length can only be constrained in uint16 type
+	*/
+	// NOTE may want to change back to uint32/uint64
+	staticLength uint16
+}
+
+// String serialize an ABI Type to a string in ABI encoding.
+func (t Type) String() string {
+	switch t.abiTypeID {
+	case Uint:
+		return "uint" + strconv.Itoa(int(t.bitSize))
+	case Byte:
+		return "byte"
+	case Ufixed:
+		return "ufixed" + strconv.Itoa(int(t.bitSize)) + "x" + strconv.Itoa(int(t.precision))
```

Comment:
No need to change, but I `fmt.Sprintf("ufixed%dx%d", t.bitSize, t.precision)` might be easier on the eyes.

Reply:
Modified.